File: frontend/lib/tasks.ts
import { decodeLease } from "./readiness-codecs.ts";
import { exactKeys, finiteInteger, objectValue, sameUuid, validUtcDateTime, validUuid } from "./wire-guards.ts";
import type { LeasePublic } from "./types.ts";

export const taskStatusLabels = {
  pending: "Pending", active: "Active", dropped: "Dropped", deferred: "Deferred",
  done: "Done", "wont-do": "Won’t do", promoted: "Promoted", superseded: "Superseded", all: "All"
} as const;
export type TaskStatus = Exclude<keyof typeof taskStatusLabels, "all">;
export type TaskKind = "work_item" | "code_review";
export type Task = {
  id: string; kind: TaskKind; project_id: string; work_item_id: string; work_version: number;
  review_state: "requested" | "completed" | "superseded" | null;
  title: string; summary: string; status: TaskStatus; updated_at: string; lease: LeasePublic | null;
};
export type TaskCounts = { active: number; pending: number };
export type TaskPage = {
  project_id: string; work_items: TaskCounts; code_reviews: TaskCounts;
  next_lease_expires_at: string | null;
  items: Task[]; total: number; limit: number; offset: number;
};
export const TASK_PAGE_SIZE = 20;

export function taskPath(task: Pick<Task, "kind" | "id" | "project_id" | "work_item_id">): string {
  const query = new URLSearchParams({ project: task.project_id, work: task.work_item_id });
  if (task.kind === "code_review") query.set("review", task.id);
  return `/${task.kind === "work_item" ? "work-items" : "code-reviews"}?${query}`;
}

export function decodeTaskPage(value: unknown, projectId: string, status: TaskStatus | "all", kind: TaskKind | undefined, offset: number): TaskPage {
  const page = objectValue(value);
  const counts = (value: unknown) => {
    const row = objectValue(value);
    return row && exactKeys(row, ["active", "pending"]) && finiteInteger(row.active) && finiteInteger(row.pending);
  };
  if (!page || !exactKeys(page, ["project_id", "work_items", "code_reviews", "next_lease_expires_at", "items", "total", "limit", "offset"])
    || !sameUuid(page.project_id, projectId) || !counts(page.work_items) || !counts(page.code_reviews)
    || page.next_lease_expires_at !== null && !validUtcDateTime(page.next_lease_expires_at)
    || !finiteInteger(page.total) || page.limit !== TASK_PAGE_SIZE || page.offset !== offset
    || !Array.isArray(page.items) || page.items.length !== Math.min(TASK_PAGE_SIZE, Math.max(0, page.total - offset))) {
    throw new Error("Mnemonic returned an invalid task page.");
  }
  const seen = new Set<string>();
  const items = page.items.map((value) => {
    const row = objectValue(value);
    if (!row || !exactKeys(row, ["id", "kind", "project_id", "work_item_id", "work_version", "title", "summary", "status", "review_state", "updated_at", "lease"])
      || !finiteInteger(row.work_version, 1) || !validUuid(row.id) || !validUuid(row.work_item_id) || !sameUuid(row.project_id, projectId)
      || !["work_item", "code_review"].includes(String(row.kind)) || kind && row.kind !== kind
      || (row.kind === "work_item" ? row.review_state !== null : !["requested", "completed", "superseded"].includes(String(row.review_state)))
      || typeof row.title !== "string" || !row.title.trim() || typeof row.summary !== "string"
      || !Object.hasOwn(taskStatusLabels, String(row.status)) || row.status === "all"
      || status !== "all" && row.status !== status || !validUtcDateTime(row.updated_at)
      || row.kind === "work_item" && !sameUuid(row.id, row.work_item_id) || seen.has(String(row.id))) {
      throw new Error("Mnemonic returned an invalid task.");
    }
    seen.add(String(row.id));
    const lease = decodeLease(row.lease);
    if ((row.status === "active") !== Boolean(lease)
      || lease && (row.kind === "code_review" ? !sameUuid(lease.code_review_id, row.id) : lease.purpose === "code_review")) {
      throw new Error("Mnemonic returned an invalid task lease.");
    }
    return { ...row, lease } as Task;
  });
  return { ...page, items } as TaskPage;
}

File: backend/src/mnemonic_api/services/tasks.py
"""Coherent, paginated dashboard task discovery without lifecycle overlays."""

from datetime import datetime
from uuid import UUID

from sqlalchemy import and_, case, func, literal, select, union_all
from sqlalchemy.orm import Session

from mnemonic_api.models import CodeReview, WorkDuplicateMerge, WorkItem, WorkLease
from mnemonic_api.schemas import LeasePublic
from mnemonic_api.services.work_events import database_now
from mnemonic_api.services.work_items import require_project
from mnemonic_api.task_schemas import TaskCounts, TaskListQuery, TaskPage, TaskSummary


def task_rows(project_id: UUID, now: datetime):
    visible = (
        WorkItem.project_id == project_id,
        WorkItem.deleted_at.is_(None),
        WorkItem.id.not_in(select(WorkDuplicateMerge.source_work_item_id)),
    )
    implementation_lease = and_(
        WorkItem.status == "pending", WorkLease.purpose == "implementation",
    )
    work_status = case(
        (and_(implementation_lease, WorkLease.expires_at > now), "active"),
        (implementation_lease, "dropped"),
        else_=WorkItem.status,
    )
    disposition = func.coalesce(CodeReview.human_decisions[-1]["status"].astext, "to-review")
    review_lease = and_(
        CodeReview.state == "requested", disposition == "to-review",
        WorkLease.purpose == "code_review", WorkLease.code_review_id == CodeReview.id,
        WorkLease.expires_at > now,
    )
    review_status = case(
        (CodeReview.state == "superseded", "superseded"),
        (CodeReview.state == "completed", "done"),
        (disposition != "to-review", disposition),
        (review_lease, "active"),
        else_="pending",
    )

    def columns(identity, kind, status, active):
        return (
            identity.label("id"), literal(kind).label("kind"), WorkItem.project_id,
            WorkItem.id.label("work_item_id"), WorkItem.title, WorkItem.summary,
            status.label("status"), WorkItem.updated_at,
            case((active, WorkLease.work_item_id)).label("lease_work_id"),
        )

    work = select(*columns(
        WorkItem.id, "work_item", work_status,
        and_(implementation_lease, WorkLease.expires_at > now),
    )).outerjoin(WorkLease, WorkLease.work_item_id == WorkItem.id).where(*visible)
    reviews = select(*columns(
        CodeReview.id, "code_review", review_status, review_lease,
    )).select_from(CodeReview).join(WorkItem, WorkItem.id == CodeReview.work_item_id).outerjoin(
        WorkLease, WorkLease.work_item_id == WorkItem.id,
    ).where(*visible)
    return union_all(work, reviews).subquery()


def task_page(database: Session, project_id: UUID, filters: TaskListQuery) -> TaskPage:
    require_project(database, project_id)
    tasks = task_rows(project_id, database_now(database))
    counts = {(kind, status): count for kind, status, count in database.execute(
        select(tasks.c.kind, tasks.c.status, func.count()).group_by(tasks.c.kind, tasks.c.status),
    )}
    next_expiry = database.scalar(select(func.min(WorkLease.expires_at)).where(
        WorkLease.work_item_id.in_(select(tasks.c.lease_work_id)),
    ))
    conditions = []
    if filters.task_id is not None:
        conditions.append(tasks.c.id == filters.task_id)
    if filters.kind is not None:
        conditions.append(tasks.c.kind == filters.kind)
    if filters.status != "all":
        conditions.append(tasks.c.status == filters.status)
    total = database.scalar(select(func.count()).select_from(tasks).where(*conditions)) or 0
    rows = database.execute(
        select(tasks).where(*conditions).order_by(tasks.c.updated_at.desc(), tasks.c.id)
        .limit(filters.limit).offset(filters.offset),
    ).mappings().all()
    lease_ids = [row["lease_work_id"] for row in rows if row["lease_work_id"]]
    leases = {lease.work_item_id: LeasePublic.model_validate(lease) for lease in database.scalars(
        select(WorkLease).where(WorkLease.work_item_id.in_(lease_ids)),
    )} if lease_ids else {}
    items = []
    for row in rows:
        values = dict(row)
        values["lease"] = leases.get(values.pop("lease_work_id"))
        items.append(TaskSummary.model_validate(values))
    return TaskPage(
        project_id=project_id,
        work_items=TaskCounts(active=counts.get(("work_item", "active"), 0),
                              pending=counts.get(("work_item", "pending"), 0)),
        code_reviews=TaskCounts(active=counts.get(("code_review", "active"), 0),
                                pending=counts.get(("code_review", "pending"), 0)),
        next_lease_expires_at=next_expiry,
        items=items, total=total, limit=filters.limit, offset=filters.offset,
    )
